scripts/zones/Bibiki_Bay/mobs/Shen.lua

```diff
@@ -29,44 +30,71 @@ local function exitShell(mob)
+    mob:setLocalVar('inShell', 0)
 end
 
 entity.onMobInitialize = function(mob)
     mob:setMobMod(xi.mobMod.IDLE_DESPAWN, 180)
 end
 
 entity.onMobSpawn = function(mob)
+    mob:addImmunity(xi.immunity.GRAVITY)
+    mob:addImmunity(xi.immunity.BIND)
+    mob:addImmunity(xi.immunity.BLIND)
+    mob:addImmunity(xi.immunity.LIGHT_SLEEP)
+    mob:addImmunity(xi.immunity.DARK_SLEEP)
+    mob:addImmunity(xi.immunity.POISON)
+    mob:addImmunity(xi.immunity.SILENCE)
+    mob:addImmunity(xi.immunity.PARALYZE)
+    mob:addImmunity(xi.immunity.REQUIEM)
+    mob:setMod(xi.mod.WATER_ABSORB, 100)
+
     mob:setLocalVar('shellTimer', os.time() + 60)
     mob:setLocalVar('petCooldown', os.time() + 20)
     exitShell(mob)
 
     mob:addListener('MAGIC_STATE_EXIT', 'SHEN_MAGIC_EXIT', function(shen, spell)
         if spell:getID() == xi.magic.spell.FLOOD then
             mob:setMagicCastingEnabled(true)
+            -- need to remove chainspell manually so it can be done silently
+            if mob:hasStatusEffect(xi.effect.CHAINSPELL) then
+                mob:delStatusEffectSilent(xi.effect.CHAINSPELL)
+            end
         end
     end)
 end
 
 entity.onMobFight = function(mob, target)
+    -- check timer for going into or out of shell
+    if os.time() > mob:getLocalVar('shellTimer') then
+        if mob:getLocalVar('inShell') == 0 and mob:getAnimationSub() == 0 then
+            enterShell(mob)
+            mob:setLocalVar('shellTimer', os.time() + math.random(30, 100))
+        elseif mob:getLocalVar('inShell') == 1 and mob:getAnimationSub() == 1 then
+            exitShell(mob)
+            mob:setLocalVar('shellTimer', os.time() + math.random(30, 100))
+        end
+    end
+
     local mobId = mob:getID()
     local petOne = GetMobByID(mobId + 1)
     local petTwo = GetMobByID(mobId + 2)
     local petCooldown = mob:getLocalVar('petCooldown')
-    local inShell = mob:getLocalVar('inShell')
+    local mobAction = mob:getCurrentAction()
+    local mobIsBusy = false
 
-    local timeToShell = mob:getLocalVar('shellTimer')
-    if os.time() > timeToShell then
-        mob:setLocalVar('inShell', 1)
-        if inShell == 1 and mob:getAnimationSub() == 0 then
-            enterShell(mob)
-
-            local timeInShell = math.random(50, 70)
-            mob:timer(timeInShell * 1000, function(shen)
-                exitShell(mob)
-                shen:setLocalVar('shellTimer', os.time() + 60)
-                shen:setLocalVar('inShell', 0)
-            end)
-        end
+    if
+        mobAction == xi.act.MOBABILITY_START or
+        mobAction == xi.act.MOBABILITY_USING or
+        mobAction == xi.act.MOBABILITY_INTERRUPT or
+        mobAction == xi.act.MOBABILITY_FINISH or
+        mobAction == xi.act.MAGIC_START or
+        mobAction == xi.act.MAGIC_CASTING or
+        mobAction == xi.act.MAGIC_INTERRUPT or
+        mobAction == xi.act.MAGIC_FINISH or
+        not mob:actionQueueEmpty()
+    then
+        mobIsBusy = true
     end
```

Comment:
You know, I'm pretty sure we have something similar somehwere. And if not, we should definately have this globalized.
Was it with those flying snakes, like tiamat and Jorm? idr

I mean, I'm sure we WANT to easily check if a mob is busy for many mobs, without having to define this logic every time, so, if you could add a TODO comment at the very least, that would help

Reply:
> You know, I'm pretty sure we have something similar somehwere. And if not, we should definately have this globalized. Was it with those flying snakes, like tiamat and Jorm? idr
> 
> I mean, I'm sure we WANT to easily check if a mob is busy for many mobs, without having to define this logic every time, so, if you could add a TODO comment at the very least, that would help

Thanks, I have updated the PR with a TODO comment. I will look into a global function at some point and put in a separate PR.